### PR TITLE
Fix #69 and first step to decouple media processing actions from episode handling

### LINKED ISSUE #69: Error while adding new show
Cannot add new show, probably since here: https://github.com/mmmnmnm/lahmacun_arcsi/commit/45594d38e60f9006c10858b2f9894f240f861fcd#diff-07f44b067558db42861b61488f3f0d808b24cadc9e6208a650b5242c35dd328b

@pvj can you check? Here the trace:

image

### PR DESCRIPTION
**Pull request by pvj**
**Why?**
It started off as quick-fix for issue #69 but since we had issues with this weak setup we used anyway I started to separate media processing, media archiving, media broadcasting. It is not full rewrite but just step in that direction while the app still works with the same UI and API payload.

**What did I do?**
Introduced `archive()`, `process()` generic functions instead of `process_media()` interface. 

Rewrote parts of add show, edit show, add item, edit item where we deal w/ media handling to use these functions. I made small changes in `form_filename()` (*this is the relevant part to issue #69 fixing that!*) and attempted to improve `broadcast_audio()` too. 

I also formatted the files I touched w/ black. We lost that somewhere but I think it is really-really good and well opinionated so I hope we keep using it.

**Does this do what it is intended to do and nothing more?**
I did tests to verify that this refactor doesn't cause any regressions. 

Add new show 
 - w/ image
   - expected: New show created
   - result: ✅ 
 - w/o image 
   - expected: New show created
   - result: ✅ 

Edit show 
 - w/ image
   - expected: Show cover updated
   - result: ✅ 
 - w/o image 
   - expected: No changes
   - result: ✅ 

Add new episode
 - archive
   - w/o image or audio
     - expected: New ep. created
     - result: ✅ 
   - w/ image only
     - expected: New ep. created
     - result: ✅ 
   - w/ audio only
     - expected: New ep. created
     - result: ✅ 
   - w/ image & audio
     - expected: New ep. created
     - result: ✅ 
 - broadcast
   - w/o image or audio
     - expected: Returns error
     - result: ✅ 
   - w/ image only
     - expected: Returns error
     - result: ✅ 
   - w/ audio only
     - expected: Returns error
     - result: ✅ 
   - w/ image & audio
     - expected: New ep. broadcasted
     - result: ✅ 
    
Edit episode
 - archive
   - w/o image or audio
     - expected: New ep. created
     - result: ✅ 
   - w/ image only
     - expected: New ep. created
     - result: ✅ 
   - w/ audio only
     - expected: New ep. created
     - result: ✅ 
   - w/ image & audio
     - expected: New ep. created
     - result: ✅ 
 - broadcast
   - w/o image or audio
     - expected: Returns error
     - result: ✅ 
   - w/ image only
     - expected: Returns error
     - result: ✅ 
   - w/ audio only
     - expected: Returns error
     - result: ✅ 
   - w/ image & audio
     - expected: New ep. broadcasted
     - result: ✅ 

Sorry for the bad branch name and introducing massive changes. This is just the first step. My code is so bad, I'll need to rewrite it quite heavily until it is effective and simple. 

Please let me know if you find something. I didn't have much time to really flesh this out but I would like to have this available for handling upcoming new shows.

### CODE REVIEW

File: arcsi/api/utils.py [hunk 1]
```diff
@@ -29,127 +29,88 @@ def media_path(show, number, item_name):
...
 
-def form_filename(file_obj, item_obj):
-    '''
-    Filename naming schema: 
-    Use normalised combination of show name and episode name of parent object. 
+def show_or_not(item_obj):
+    return True if isinstance(item_obj, Show) else False
+
+
+def slug(namestring):
```
Comment (gammaw) [inline arcsi/api/utils.py:42]: Couldn't find where you use it?

File: arcsi/api/utils.py [hunk 2]
```diff
@@ -29,127 +29,88 @@ def media_path(show, number, item_name):
...
     Join them w/ delimiter.
     Get the extension from file sent to API.
     To get the extension we use rsplit w/ maxsplit=1 to make sure we always get the extension even if there is another dot in the filename.
-    '''
+    """
     ext = file_obj.filename.rsplit(DOT, 1)[1].lower()
-    norms_show_name = normalise(item_obj.shows[0].name)
-    norms_ep_name = normalise(item_obj.name)
+    norms_show_name = normalise(title_tuple[0])
+    norms_ep_name = normalise(title_tuple[1])
```
Comment (gammaw) [inline arcsi/api/utils.py:57]: I see the trick! Good job!

**Review by gammaw — APPROVED** [review]
Great job! Did a quick code review, got the very idea I guess (your intro text helped a lot), makes sense to me. I didn't try it, since you did a thorough functional testing. Added some comments, mostly cosmetics & optimisations. Disclaimer: I wasn't the freshest when I did all this.

**Comment by gammaw** [comment]
One more question: 

> I also formatted the files I touched w/ black. We lost that somewhere but I think it is really-really good and well opinionated so I hope we keep using it.

Sorry, I didn't get this one. "w/ black"?

File: arcsi/api/item.py [hunk 3]
```diff
@@ -141,8 +156,69 @@ def add_item():
...
+                if request.files["play_file"] != "":
+                    play_file = request.files["play_file"]
+                    new_item.play_file_name = process(
+                        archive_base=new_item.shows[0].archive_lahmastore_base_url,
+                        archive_idx=new_item.number,
+                        archive_file=play_file,
+                        archive_name=(new_item.shows[0].name, new_item.name),
+                    )
+            
+            if not (image_file_name or new_item.play_file_name):
```
[the remark this answers is not in the document]
Comment (pvj) [reply arcsi/api/item.py:181]: The current idea we agreed is that we may have episodes w/o media. So this is why we won't block this right at the beginning. But yeah I deliberated this many times too. It's not a big overhead right now so why not let us post plain records to persist.

File: arcsi/api/utils.py [hunk 4]
```diff
@@ -29,127 +29,88 @@ def media_path(show, number, item_name):
...
+        broadcast_title,
+        broadcast_playlist,
     )
 
     # TODO find image -- fallback to show cover; handle this if-tree better
     # TODO embed metadata regardless of there's image or not f.e. title && artist
     # file_path but then image_url?? ugly
-    if item.image_url:
+    if image_file_path:
         episode_update = az.embedded_metadata()
```
[the remark this answers is not in the document]
Comment (pvj) [reply arcsi/api/utils.py:86]: Yeah this is one of the next steps and should be part of processing!

File: arcsi/api/utils.py (hunk 1 above)
Comment (pvj) [reply arcsi/api/utils.py:42]: Yeah this is a leftover. Good catch!

**Comment by pvj** [comment]
> One more question:
> 
> > I also formatted the files I touched w/ black. We lost that somewhere but I think it is really-really good and well opinionated so I hope we keep using it.
> 
> Sorry, I didn't get this one. "w/ black"?

It's an opiniated python code formatter that enforces some style guide rules for sake of ease & sanity. https://github.com/psf/black

**Pull request merged**